File: api/scalars/slug.py
from re import compile
from graphene.types import Scalar
from graphql.language import ast
from graphql import GraphQLError

from functions.error_messages import *

SLUG_REGEX = r"""^[a-z0-9]+(?:-[a-z0-9]+)*$"""
SLUG_REGEX = compile(SLUG_REGEX)


class Slug(Scalar):
    """
    A field whos values contain numbers, letters, dashes, and underscores
    """

    @staticmethod
    def serialize(value):
        if not isinstance(value, str):
            raise GraphQLError(scalar_error_type("String", value))

        if not SLUG_REGEX.search(value):
            raise GraphQLError(scalar_error_type("Slug", value))
        return value

    @staticmethod
    def parse_value(value):
        if not isinstance(value, str):
            raise GraphQLError(scalar_error_type("String", value))

        if not SLUG_REGEX.search(value):
            raise GraphQLError(scalar_error_type("Slug", value))
        return value

    @staticmethod
    def parse_literal(node):
        if not isinstance(node, ast.StringValue):
            raise GraphQLError(
                scalar_error_only_types("Strings", "Slug", str(ast.Type))
            )

        if not SLUG_REGEX.search(node.value):
            raise GraphQLError(scalar_error_type("Slug", node.value))
        return node.value

File: api/src/scalars/slug.js
import { Kind, GraphQLError, GraphQLScalarType } from 'graphql'

const validate = (value) => {
  const SLUG_REGEX = /^[a-z0-9]+(?:-[a-z0-9]+)*$/

  if (typeof value !== 'string') {
    throw new TypeError(`Value is not string: ${typeof value}`)
  }

  if (!SLUG_REGEX.test(value)) {
    throw new TypeError(`Value is not a valid slug: ${value}`)
  }
  return value
}

export const Slug = new GraphQLScalarType({
  name: 'Slug',
  description:
    'A field whose values contain numbers, letters, dashes, and underscores.',
  serialize: validate,
  parseValue: validate,

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Can only validate strings as slug but got a: ${ast.kind}`,
      )
    }
    return validate(ast.value)
  },
})
